odxtools/diaglayerraw.py

```diff
@@ -0,0 +1,283 @@
+            for el in et_element.iterfind("IMPORT-REFS/IMPORT-REF")
+        ]
+
+        state_charts = [
+            StateChart.from_et(el, doc_frags)
+            for el in et_element.iterfind("STATE-CHARTS/STATE-CHART")
+        ]
+
+        additional_audiences = [
+            AdditionalAudience.from_et(el, doc_frags)
+            for el in et_element.iterfind("ADDITIONAL-AUDIENCES/ADDITIONAL-AUDIENCE")
+        ]
+
+        sdgs = create_sdgs_from_et(et_element.find("SDGS"), doc_frags)
+
+        parent_refs = [
+            ParentRef.from_et(pr_el, doc_frags)
+            for pr_el in et_element.iterfind("PARENT-REFS/PARENT-REF")
+        ]
+
+        communication_parameters = [
+            CommunicationParameterRef.from_et(el, doc_frags, variant_type)
+            for el in et_element.iterfind("COMPARAM-REFS/COMPARAM-REF")
+        ]
+
+        ecu_variant_patterns = [
+            EcuVariantPattern.from_et(el, doc_frags)
+            for el in et_element.iterfind("ECU-VARIANT-PATTERN/ECU-VARIANT-PATTERN")
+        ]
+        if variant_type is not DiagLayerType.ECU_VARIANT:
+            assert (
+                len(ecu_variant_patterns) == 0
+            ), "DiagLayer of type other than 'ECU-VARIANT' must not define a ECU-VARIANT-PATTERN"
+
+        # Create DiagLayer
+        return DiagLayerRaw(
+            variant_type=variant_type,
+            odx_id=odx_id,
+            short_name=short_name,
+            long_name=long_name,
+            description=description,
+            admin_data=admin_data,
+            company_datas=NamedItemList(short_name_as_id, company_datas),
+            functional_classes=NamedItemList(short_name_as_id, functional_classes),
+            diag_data_dictionary_spec=diag_data_dictionary_spec,
+            diag_comms=diag_comms,
+            requests=NamedItemList(short_name_as_id, requests),
+            positive_responses=NamedItemList(short_name_as_id, positive_responses),
+            negative_responses=NamedItemList(short_name_as_id, negative_responses),
+            global_negative_responses=NamedItemList(short_name_as_id, global_negative_responses),
+            import_refs=import_refs,
+            state_charts=NamedItemList(short_name_as_id, state_charts),
+            additional_audiences=NamedItemList(short_name_as_id, additional_audiences),
+            sdgs=sdgs,
+            parent_refs=parent_refs,
+            communication_parameters=communication_parameters,
+            ecu_variant_patterns=ecu_variant_patterns,
+        )
+
+    def _build_odxlinks(self) -> Dict[OdxLinkId, Any]:
+        """Construct a mapping from IDs to all objects that are contained in this diagnostic layer."""
+        odxlinks = {self.odx_id: self}
+
+        for obj in chain(
+            [self.admin_data],
+                self.company_datas,
+                self.functional_classes,
+            [self.diag_data_dictionary_spec],
+                self.diag_comms,
```

Comment:
Indent does not look right here

Reply:
I also noticed that. The problem is that this seems to be a bug in our style police (YaPF), and CI fails if YaPF reformats anything. (I would like to keep it this way to avoid larger formatting issues...)

I concluded that we have to live with this issue for now until it is fixed in YaPF...